### ISSUE TITLE
Warn or error if non-keyword strings having "@" are encountered

### ISSUE THREAD

**Issue by gkellogg** [w3c/json-ld-syntax#16]
Looking at a [schema.org example for ItemList](http://schema.org/ItemList#ItemList-gen-306), there is a JSON-LD example which includes the fictitious `@url` keyword, where they likely meant `@id`.

```jsonld
{
  "@context": "http://schema.org",
  "@type": "ItemList",
  "@url": "http://en.wikipedia.org/wiki/Billboard_200",
  "name": "Top music artists",
  "description": "The artists with the most cumulative weeks at number one according to Billboard 200",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@type": "MusicGroup",
        "name": "Beatles"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@type": "MusicGroup",
        "name": "Elvis Presley"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@type": "MusicGroup",
        "name": "Michael Jackson"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@type": "MusicGroup",
        "name": "Garth Brooks"
      }
    }
  ]
}
```

There's nothing to signal an issue, and a JSON-LD processor will happily resolve this relative to `@vocab` as `http://schema.org/@url`. This creates a potential forward-compatibility issue if new keywords are introduced, as they are in 1.1. We might want to describe normative or suggested behavior if a processor encounters a string which could hold a keyword, but holds something else starting with `@`.

cc/ @danbri

Original issue: [Warn or error if non-keyword strings having "@" are encountered #598](https://github.com/json-ld/json-ld.org/issues/598)

**Comment by ajs6f** [comment]
Is it possible to choose to ignore such new `@` keywords in the algorithms? I.e. create no triples from them and drop them in expansion?

**Comment by gkellogg** [comment]
> Is it possible to choose to ignore such new @ keywords in the algorithms? I.e. create no triples from them and drop them in expansion?

That's what 1.0 does, it just ignores them. The issue was an example in schema.org ([ItemList example #3](https://schema.org/ItemList#ItemList-gen-306)) which uses `@url` instead of `@id`, which would seem to be an error. Thus the danger of ignoring such things.

```{
  "@context": "http://schema.org",
  "@type": "ItemList",
  "@url": "http://en.wikipedia.org/wiki/Billboard_200",
  "name": "Top music artists",
  "description": "The artists with the most cumulative weeks at number one according to Billboard 200",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@type": "MusicGroup",
        "name": "Beatles"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@type": "MusicGroup",
        "name": "Elvis Presley"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@type": "MusicGroup",
        "name": "Michael Jackson"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@type": "MusicGroup",
        "name": "Garth Brooks"
      }
    }
  ]
}
```

**Comment by ajs6f** [comment]
I'm not sure if I'm misunderstanding, but the example doesn't seem to have ignored it at all-- it expanded it to `http://schema.org/@url`. I'm asking whether it would be possible to ignore it in expansion entirely, i.e. `@url` would be dropped entirely, no expanded form, no triples. That might be more surprising than helpful, but it would seem to be a little more forward-compatible. Or perhaps I'm still missing the point (wouldn't be the first time!)?

**Comment by gkellogg** [comment]
> the example doesn't seem to have ignored it at all

Good point, that's because of the use of `@vocab`; otherwise, it would have been dropped. Another reason we might flag this. If you really wanted `@url` as a relative IRI, you could always use `./@url`.

**Comment by dlongley** [comment]
This seems like a feature implementations could offer (a special mode to check for this sort of thing). I don't think we'd want this extra burden (or any similar extra burdens) added to the standard processing rules.

**Comment by ajs6f** [comment]
@dlongley I don't maintain an impl (so take a grain of salt with this question), but would  a configurable mode just shift the burden to the clients of impls instead of the maintainers of impls?

**Comment by dlongley** [comment]
@ajs6f,

Yes, but I think that's where it should be. In fact, I think I'm speaking more with my client hat on than my implementer hat when I say that I do not want my production systems to have to deal with the extra burden of checking non-error cases to express warnings.

I think it should be a goal to have a "fast lane" for JSON-LD processing to keep things as efficient as possible. I would prefer that the bottleneck in a complex software system never be JSON-LD processing. One way to help support that effort is to keep corner cases like raising warnings out of the standard processing mode.

**Comment by azaroth42** [comment]
I'm convinced by @dlongley's argument. 

I'm in favor of close won't fix for the following rationale:

* Per @dlongley, raising a warning but otherwise continuing is a burden for processing that in the regular case of it being a part of a much larger automated system will never be seen anyway, so just adds complexity and no value.
* It could be an option to turn on when debugging your JSON-LD, and turned on by default in the playgrounds. But that's an implementation issue, not syntax or API.
* It would break the (admittedly incorrect) existing usage in the schema example, but I wouldn't be surprised if someone had intentionally mapped `@label` to `rdf:label` to go along with `@type`, and it would certainly break them.
* There is already a non-normative note at the end of 4.8 warning against making your own keys that start with `@`, and I think from a spec perspective that is the right approach.

**Comment by ajs6f** [comment]
@azaroth42 I'm actually fine with close-won't-fix, but I do want to address one point you make:
> Per @dlongley, raising a warning but otherwise continuing is a burden for processing that in the regular case of it being a part of a much larger automated system will never be seen anyway, so just adds complexity and no value.

I disagree _strongly_. It's exactly in the case of a large fully-automated system where I _want_ warnings from my libraries (instead of silent acquiescence), so that I can accumulate information that lets me improve workflows without them having to break down first. So although I agree that this ticket isn't a good case at all for emitting warnings, I disagree strongly that we could find _no_ such cases.

Just my 2¢!

**Comment by azaroth42** [comment]
WG [resolved](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2018/2018-08-24-json-ld#resolution2) to add a non-normative note to the API specification to recommend that implementations generate a warning, however is appropriate for the implementation's language, when `@` signs are encountered outside of specified keywords.

**Comment by gkellogg** [comment]
Fixed in 760b32fd396e47f5ae0ff720c3f3c9a2080592eb.

**Comment by azaroth42** [comment]
Reopening from TPAC discussion

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-09-20-json-ld#section5-2).

- `RESOLVED:  Reopen w3c/json-ld-syntax#16 and require processors to ignore terms beginning with @ that are not defined keywords`
<details><summary><i class="differentiate">View the transcript</i></summary>
Ivan Herman: See Issue Syntax#260<br/>
Gregg Kellogg: what we’re most worried about is other specs using @ and we want them to be compatible with us<br/>
… but at this point hopefully everyone’s aware of our use of @<br/>
Manu Sporny: what does changing it to MUST NOT buy us?<br/>
Gregg Kellogg: it means that if we add things like <code>@direction</code> in the future, the processors would choke on the new term but they also wouldn’t try to turn it into a URI<br/>
Ralph Swick: so, could you just say SHOULD NOT?<br/>
Rob Sanderson: it does<br/>
Ralph Swick: ah, so you are leaving this open for future compatibility<br/>
Gregg Kellogg: and I think <code>@direction</code> is an example of why we want this<br/>
Rob Sanderson: there was an earlier issue which we closed about this<br/>
… and it was mentioned that if we found new info we might reopen it<br/>
… this seems like new infromation<br/>
Gregg Kellogg: so, some future <code>@direction</code> type thing could be setup to just disappear<br/>
Rob Sanderson: or attempt to turn into a crazy relative IRI<br/>
Benjamin Young: so, as long as we just ignore the new values when processing–and don’t choke–that should leave the door open to the future without tripping up old clients<br/>
Gregg Kellogg: authors should not use these keywords, but processors must ignore them<br/>
… but with a warning<br/>
Proposed resolution: Reopen w3c/json-ld-syntax#16 and require processors to ignore terms beginning with @ that are not defined keywords (Rob Sanderson)<br/>
Ivan Herman: See Issue syntax#16<br/>
Gregg Kellogg: +1<br/>
Manu Sporny: +1<br/>
Rob Sanderson: +1<br/>
Benjamin Young: +1<br/>
Simon Steyskal: +1<br/>
Pierre-Antoine Champin: +1<br/>
Ivan Herman: +1<br/>
Resolution #5: Reopen w3c/json-ld-syntax#16 and require processors to ignore terms beginning with @ that are not defined keywords<br/>
Rob Sanderson: Have we made this clear, it’s resolved?<br/>
Gregg Kellogg: Yes, the editors need to do their thing now.<br/>
Rob Sanderson: so we’ll close #260 in favor of opening #16.<br/>
</details>
